`storage`: check if segments need rewrite in sliding window compaction

WillemKauf opened the pull request:
Segments considered during sliding window compaction need rewriting _only_ iff they contain at least one key found in the `key_offset_map` which is _not_ the latest offset for that key.

In an attempt to reduce unnecessary CPU and IO work effectively copying the segment along with the compacted index and segment index to new files, we can use a compacted index reader to determine whether the segment actually requires a rewrite or not before doing so.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* Reduce unnecessary CPU & IO work during compaction by avoiding rewrites of `segment` files that would remove no data (effectively be copies).

WillemKauf committed:
`storage`: add `segment_needs_rewrite_with_offset_map()`

Segments considered during sliding window compaction need rewriting _only_
iff they contain at least one key found in the `key_offset_map` which is
_not_ the latest offset for that key.

In an attempt to reduce unnecessary CPU and IO work effectively copying the
segment along with the compacted index and segment index to new files,
we can use a compacted index reader to determine whether the segment actually
requires a rewrite or not before doing so.

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/bfa61acc-e5f7-4406-a0cf-62fd233ffb9b?refs=7a1da46599a8876fd8d3ce3e8a9a924125272df9&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/bfa61acc-e5f7-4406-a0cf-62fd233ffb9b/license?refs=7a1da46599a8876fd8d3ce3e8a9a924125272df9&source=prChecksComment)

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/1b8e2c51-43b0-4cc1-8c7e-b63e783da671?refs=7a1da46599a8876fd8d3ce3e8a9a924125272df9&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/1b8e2c51-43b0-4cc1-8c7e-b63e783da671/license?refs=7a1da46599a8876fd8d3ce3e8a9a924125272df9&source=prChecksComment)

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65078-->
<!--2322e1f7-f648-42f4-881e-af8de3b694e8-->
<!--2cf41eee-3eb1-42fa-9b24-f34bdb09d9db-->
<!--title-->
## Retry command for [Build#65078](https://buildkite.com/redpanda/redpanda/builds/65078)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":true}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":false}
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d03a86ec-5429-42cb-a937-06e70993f615-->
<!--TEST RESULTS-->
<!--62ba72fc-fc33-4b71-b2b5-e4090e99929b-->
<!--TEST RESULTS-->
<!--0acaee8b-a4e0-47f1-aecb-b62d17a862d4-->
<!--TEST RESULTS-->
<!--be11d0d7-5019-4f53-befb-57aa231e39a3-->
<!--TEST RESULTS-->
<!--0501c75d-c7c5-4937-aa98-e821ffb4067c-->
<!--TEST RESULTS-->
<!--f10e645f-9c5c-4055-8cb0-9f949beeed36-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65078">build#65078</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65078#01966b09-0c2f-461b-ac87-ce31db42c3a3 | FLAKY | 1/2 |
| opfuzz_test_rpunit.opfuzz_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65078#01966b09-0c2f-461b-ac87-ce31db42c3a3 | FAIL | 0/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/65078#01966b47-e63f-46ba-8355-30d32eed7a9f | FLAKY | 16/21 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/65078#01966b47-e63d-4a7b-8661-437fbdb416a3 | FLAKY | 15/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/65078#01966b47-e63f-46ba-8355-30d32eed7a9f | FLAKY | 9/21 |
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65078#01966b09-0c2f-461b-ac87-ce31db42c3a3 | FAIL | 0/2 |
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65078#01966b09-0c30-40d7-819e-ec746cc67476 | FAIL | 0/2 |
| storage_single_thread_rpunit.storage_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65078#01966b09-0c30-40d7-819e-ec746cc67476 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65116">build#65116</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.datalake_e2e_test.DatalakeDelayedEnablementTest.test_enabling_iceberg_in_existing_cluster.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/65116#01966dde-7726-481c-9aa2-e0c22ba184e7 | FLAKY | 17/21 |
| rptest.tests.datalake.rest_catalog_connection_test.RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/65116#01966dde-7726-481c-9aa2-e0c22ba184e7 | FLAKY | 19/21 |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_rack_awareness | ducktape | https://buildkite.com/redpanda/redpanda/builds/65116#01966dde-7725-422b-a1b1-7139e1c1b4f7 | FLAKY | 20/21 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/65116#01966dde-7725-422b-a1b1-7139e1c1b4f7 | FLAKY | 20/21 |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65116#01966dde-7725-422b-a1b1-7139e1c1b4f7 | FLAKY | 14/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65205">build#65205</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65205#01967f13-e5b9-4fdf-ab5b-77be3f280aa6 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65369">build#65369</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/65369#019689ba-18d7-4c29-8b69-805404696d00 | FLAKY | 19/21 |
| rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=node_add.test_mode=TestMode.FAST_MOVES.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/65369#019689ba-18d6-4c97-aedf-19dfb0901fb9 | FLAKY | 20/21 |
| rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=off.test_mode=TestMode.TIRED_STORAGE.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/65369#019689ba-18d8-4a93-beea-e405d93c40b3 | FLAKY | 19/21 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments | ducktape | https://buildkite.com/redpanda/redpanda/builds/65369#019689ba-18d5-45c0-8076-2f8a4cb7ea4b | FLAKY | 19/21 |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65369#0196899c-4d9b-472e-80cc-57aac75fcaa5 | FLAKY | 20/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65417">build#65417</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.internal_topic_protection_test.InternalTopicProtectionLargeClusterTest.test_consumer_offset_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/65417#01968da8-464f-4b61-8635-0f3d03b99384 | FLAKY | 20/21 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=True.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/65417#01968dab-9ded-43a2-a38a-473f61bbec31 | FLAKY | 20/21 |
| rptest.tests.topic_delete_test.TopicDeleteCloudStorageTest.drop_lifecycle_marker_test.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/65417#01968da8-464f-4b61-8635-0f3d03b99384 | FLAKY | 20/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/65417#01968da8-464d-4719-bc34-5df7ae7b1299 | FLAKY | 11/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65492">build#65492</a></summary>


| test_id                                                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=100.storage_compaction_key_map_memory_kb=10 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65492#01969345-75d8-4cbc-a130-34684ff417e3 | FLAKY         | 20/21    |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_fuzz_admin_ops                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65492#01969345-75d8-4cbc-a130-34684ff417e3 | FLAKY         | 17/21    |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65492#01969345-75d9-4c50-bd5e-4ee7ce26f177 | FLAKY         | 12/21    |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65492#01969345-75d9-4c50-bd5e-4ee7ce26f177 | FLAKY         | 10/21    |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=2                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65492#01969341-1866-486c-abdf-f1824cd9dde6 | FLAKY         | 20/21    |
</details>
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65116-->
<!--6c808403-839f-47de-8bc3-6d78d7ad80df-->
<!--435f561c-dad0-4f8e-9c5e-111e886218ef-->
<!--title-->
## Retry command for [Build#65116](https://buildkite.com/redpanda/redpanda/builds/65116)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/datalake/datalake_e2e_test.py::DatalakeDelayedEnablementTest.test_enabling_iceberg_in_existing_cluster@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/datalake/rest_catalog_connection_test.py::RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog@{"cloud_storage_type":1}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":false,"mixed_versions":false,"with_chunked_compaction":true,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_rack_awareness
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
```
<!--/body-->

wdberkeley commented on src/v/storage/segment_deduplication_utils.cc:
```diff
@@ -298,4 +303,44 @@ ss::future<bool> index_chunk_of_segment_for_map(
+  const compaction_config& cfg,
+  ss::lw_shared_ptr<segment> seg,
+  const key_offset_map& map) {
+    auto compaction_idx_path = seg->path().to_compacted_index();
```
Perhaps not necessary for this change, but would it make sense to put L310-323 into a `make_compacted_index_reader` method on `segment`?

WillemKauf replied:
Could be, I think this chunk may appear in a number of places in the code, but I'll leave the addition of a new function for a future PR.

WillemKauf commented:
/ci-repeat 1
tests/rptest/tests/datalake/datalake_e2e_test.py::DatalakeDelayedEnablementTest.test_enabling_iceberg_in_existing_cluster@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/datalake/rest_catalog_connection_test.py::RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog@{"cloud_storage_type":1}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":false,"mixed_versions":false,"with_chunked_compaction":true,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_rack_awareness
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65199-->
<!--0ab40310-37ed-4ccc-841f-d179776b5091-->
<!--title-->
## Retry command for [Build#65199](https://buildkite.com/redpanda/redpanda/builds/65199)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/rest_catalog_connection_test.py::RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog@{"cloud_storage_type":1}
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
tests/rptest/tests/datalake/rest_catalog_connection_test.py::RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog@{"cloud_storage_type":1}

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -763,6 +763,23 @@ ss::future<bool> disk_log_impl::sliding_window_compact(
         const bool is_clean_compacted = seg->offsets().get_base_offset()
                                         >= idx_start_offset;
 
+        if (!co_await segment_needs_rewrite_with_offset_map(cfg, seg, map)) {
```
Do we need to account for whether the segment has tombstones that are ready to be removed? Wondering if this check could cause us to skip tombstone removal... though maybe not because I think we're guaranteed to compact the segment at least once before marking it as finished + clean.

WillemKauf replied:
I considered this, but [self compaction occurs for all segments in the sliding window](https://github.com/redpanda-data/redpanda/blob/9e9d74c9bf23b4a75a6f21cfeb693bfe18605896/src/v/storage/disk_log_impl.cc#L613-L618) before de-duplication does, and there we will forcefully self compact if there are any [removable tombstones present](https://github.com/redpanda-data/redpanda/blob/9e9d74c9bf23b4a75a6f21cfeb693bfe18605896/src/v/storage/segment_utils.cc#L747-L761).

We _could_ consider whether tombstones would be removed by the rewrite here, but it's very minor and likely that tombstones will be removed by self compaction in the future anyways. WDYT?

andrwng replied:
Ah, it's probably fine as is.

I do want to make sure that in working on longer term work to schedule compaction better that we don't miss this somewhat nuanced dependence on self compaction to satisfy tombstone removal.

WillemKauf replied:
Then again I don't think it's hard to add an `&&` condition here. I think I will.

WillemKauf replied:
Pushed.

WillemKauf committed:
`storage`: also consider removable tombstones in sliding window

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cloud-1t5/pr-checks/37b1727f-3388-4539-83b3-adba75dcb136?refs=7a1da46599a8876fd8d3ce3e8a9a924125272df9&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cloud-1t5/pr-checks/37b1727f-3388-4539-83b3-adba75dcb136/license?refs=7a1da46599a8876fd8d3ce3e8a9a924125272df9&source=prChecksComment)

WillemKauf commented:
Fun new race condition with prefix truncation I introduced:

For context, the failed test `write_truncate_compact` in `storage_e2e_test` expects that `mgr.remove(ntp).get();` should remove all persistent files on disk.

```
DEBUG 2025-04-30 23:56:01,241 [shard 0:main] storage - segment.cc:171 - removed: "test.data.VmorSj8cNX/default/test/0_0/61-0-v1.compaction_index" size 333
DEBUG 2025-04-30 23:56:01,253 [shard 0:main] storage-gc - disk_log_impl.cc:1514 - [{default/test/0}] Deduplicating data from segment test.data.VmorSj8cNX/default/test/0_0/61-0-v1.log to test.data.VmorSj8cNX/default/test/0_0/61-0-v1.log.compaction.staging: {offset_tracker:{term:0, base_offset:61, committed_offset:78, dirty_offset:78}, compacted_segment=1, finished_self_compaction=1, finished_windowed_compaction=0, generation=25, reader={test.data.VmorSj8cNX/default/test/0_0/61-0-v1.log, (1339 bytes)}, writer=nullptr, cache=nullptr, compaction_index:nullopt, closed=1, tombstone=1, index={file:test.data.VmorSj8cNX/default/test/0_0/61-0-v1.base_index, offsets:0, index:{header_bitflags:0, base_offset:0, max_offset:78, base_timestamp:{timestamp: 1746057361161}, max_timestamp:{timestamp: 1746057361161}, batch_timestamps_are_monotonic:1, with_offset:true, non_data_timestamps:0, broker_timestamp:{{timestamp: 1746057361179}}, num_compactible_records_appended:{17}, clean_compact_timestamp:{nullopt}, may_have_tombstone_records:0, index(1, 1, 1)}, step:32768, needs_persistence:0}}
INFO  2025-04-30 23:56:01,352 [shard 0:main] storage_test_fixture::write_truncate_compact - Leftover file: 61-0-v1.compaction_index
src/v/storage/tests/storage_e2e_test.cc(3089): fatal error: in "write_truncate_compact": critical check false == ss::file_exists(dir_path).get() has failed [false != true]
```

`internal::make_reader_handle(compaction_idx_path, cfg.sanitizer_config);` actually [opens a file with `ss::open_flags::ro | ss::open_flags::create`](https://github.com/redpanda-data/redpanda/blob/18b0c24d7b306add4a1390fc2c56a5de5420f567/src/v/storage/segment_utils.cc#L146) flags for some reason...

Fix the race by early exiting if the `.compaction_index` file doesn't exist.


```diff
 ss::future<bool> segment_needs_rewrite_with_offset_map(
   const compaction_config& cfg,
   ss::lw_shared_ptr<segment> seg,
   const key_offset_map& map) {
     auto compaction_idx_path = seg->path().to_compacted_index();
+    // If the file doesn't exist for whatever reason, return true.
+    // This could, for example, race with a truncation which removes the
+    // .compaction_index.
+    if (!co_await ss::file_exists(compaction_idx_path.string())) {
+        co_return true;
+    }
...
```

WillemKauf commented:
^^^^ 

The above is only possible because of where we check for closed segments in the existing compaction utils; I opened PR https://github.com/redpanda-data/redpanda/pull/26007 to improve this process :)

WillemKauf committed:
`storage`: bump Bazel timeout for `compaction_fuzz_test`

This test has timed out in CI before with the `short` timeout setting.

Bump it to moderate.

WillemKauf commented:
Push to:
* Add commit bumping `compaction_fuzz_test` timeout in Bazel to `moderate`, as it timed out in this PR's CI (though I don't think this indicates an issue with the changes in the PR).

WillemKauf commented:
Force push to:
* Rebase to `upstream/dev` to fix merge conflict

WillemKauf committed:
`storage`: check if segment needs rewrite in sliding window compaction

If rewriting the segment and its files using the constructed `key_offset_map`
would have no effect other than unnecessary copying, skip rewriting them
in sliding window compaction & chunked sliding window compaction.

WillemKauf commented:
Force push to:
* Improve boolean logic to save on calls to `segment_needs_rewrite_with_offset_map()` iff a `segment->may_have_removable_tombstones()`
